docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+and/or the request parameters to allow more control.
+
+---
+## Consequences
+
+- ✅ Reduces the number of transport options for RPC.  Supports [RFC 002][rfc002].
+- ️✅ Removes the primary non-standard use of JSON-RPC.
+- ⛔️ Forces clients to migrate to a different API (eventually).
+- ↕️  API requires clients to poll, but this reduces client state on the server.
+- ↕️  We have to maintain both implementations for a whole release, but this
+  gives clients time to migrate.
+
+---
+## Alternative Approaches
+
+The following alternative approaches were considered:
+
+1. **Leave it alone.** Since existing tools mostly already work with the API as
+   it stands today, we could leave it alone and do our best to improve its
+   performance and reliability.
+
+   Based on many issues reported by users and node operators (e.g.,
+   [#3380][i3380], [#6439][i6439], [#6729][i6729], [#7247][i7247]), the
+   problems described here affect even the existing use that works.  Investing
+   further incremental effort in the existing API is unlikely to address these
+   issues.
+
+2. **Design a better streaming API.** Instead of polling, we might try to
+   design a better "streaming" API for event subscription.
+
+   A significant advantage of switching away from streaming is to remove the
+   need for persistent connections between the node and subscribers.  A new
+   streaming protocol design would lose that advantage, and would still need a
+   way to let clients recover and replay.
+
+   This approach might look better if we decided to use a different protocol
+   for event subscription, say gRPC instead of JSON-RPC. That choice, however,
+   would be just as breaking for existing clients, for marginal benefit.
+   Moreover, this option increases both the complexity and the resource cost on
+   the node implementation.
+
+   Given that resource consumption and complexity are important considerations,
+   this option was not chosen.
+
+3. **Defer to an external event broker.** We might remove the entire event
+   subscription infrastructure from the node, and define an optional interface
+   to allow the node to publish all its events to an external event broker,
+   such as Apache Kafka.
+
+   This has the advantage of greatly simplifying the node, but at a great cost
+   to the node operator: To enable event subscription in this design, the
+   operator has to stand up and maintain a separate process in communion with
+   the node, and configuration changes would have to be coordinated across
+   both.
+
+   Moreover, this approach would be highly disruptive to existing client use,
+   and migration would probably require switching to third-party libraries.
+   Despite the potential benefits for the node itself, the costs to operators
+   and clients seems too large for this to be the best option.
+
+   Publishing to an external event broker might be a worthwhile future project,
+   if there is any demand for it.
```

Comment:
Supporting pluggable indexers should allow for other teams to build their own external event broker if they want it right?

Reply:
Not quite, at least as it is now.

The indexer doesn't receive all events, it only subscribes to block events. So even if someone plugged in a custom indexer, it wouldn't see the whole event stream.

We _could_ publish more to the indexer than we do today. The interface we use for the indexer is block/tx oriented, but we could generalize it.

What that means, however, is that event subscription can't (currently) live downstream of indexing—I actually considered that as an initial approach, though.

Reply:
I think having only block/tx oriented events should be sufficient